Votability and views for posts#index and posts#show

gingin77 opened the pull request:
**What:** A logged in user can up vote or down vote a post on either the Posts#index or #show view. Once they have cast a vote, a new vote instance is created and the button for the vote cast is no longer clickable. Instead, the user will see the 'x' button to cancel their vote is now orange (rather than grey) and is clickable. I tried to capture all of this in the feature test - `spec/features/vote_on_posts_spec.rb`

This PR also includes a bug fix related to assigning dependent-destroy to the wrong side of the associations.

The view partials (in the shared dir because I'll eventually be using the same views to allow votes on comments) may seem confusing. I tried to name the partials based on the scenario when they are used. For example, `_down_to_up.html.erb` is used when a user has already cast a down vote, so the view hosts a clickable up vote button, but just an icon for a down vote.
![votes_on_posts](https://user-images.githubusercontent.com/30031865/36491495-dbed7fae-16f8-11e8-889d-86471d7f9ade.gif)

gingin77 committed:
WIP - Add font-awesome gem to get up and down icons; add vote icons to posts#index view

gingin77 committed:
WIP - Setting up flow for voting on posts

gingin77 committed:
WIP - still working on how to display voting functions

gingin77 committed:
WIP - Still working out vote layout design and how to test

gingin77 committed:
resolve merge conflict, continue with rebase

gingin77 committed:
Votes controller - remove all merge conflic flags

gingin77 committed:
Introduce VoteHelper to Controller/Concerns
Specificy which votes resources are needed

gingin77 committed:
Ingegration test for vote buttons is working

At present, when a logged in user goes to the /posts/ page, they will see orange up and down vote buttons and a grey x next to posts they have NOT voted on. Once they vote, the greyed out ex switches to orange and is clickable... Next, need to refactor conditional on main index page to evaluate if a vote exists

gingin77 committed:
Add a method on post model to find votes by user_id for a given instance

gingin77 committed:
Enhance vote button appearance and functionality

Integration tests for adding an up vote and deleting a vote have been written and are passing
The value of a vote is evaluated in the posts/index view - if +1, the up arrow is greyed out and not attached to a form and vice versa
Next - look for ways to DRY up the code; Add edit action in controller and replace new_vote form with edit_vote form

gingin77 committed:
Add styles to vote button layout on Posts#index view

Refactor other views by looking for pieces that split nicely into partials
Update style sheet and test where needed to align with view updates

gingin77 committed:
Remove unnecessary methods from Votes Contr.; set up polyM redirect path

gingin77 committed:
Tried to see how things look when one can vote on a post on the posts#show page

See notes on votes controller create method
  - polymorphic_url(vote.voteable_type.constantize) << goes to posts#index
  - polymorphic_url(vote.voteable) << goes to post#show

gingin77 committed:
Base Vote Controller redirects on referrer path. See private method.

gingin77 committed:
In multiple views, switch names used in locals for partial to be parent rather than post, so the temps make more sense when used for comments

gingin77 committed:
Continue to split views for vote partials; Enhance vote button layout in post#show view

gingin77 committed:
Remove blank line from application controller.

gingin77 committed:
Move VoteHelper module to lib and set up load path in application.rb

In Votes controller, use redirect_back with root as fallback, in place
of request_referrer private method

gingin77 committed:
Resolved merge conflict for spec/models/voter_spec.rb during rebase

gingin77 committed:
Fix placement of dependent destroy

After rebasing, I found a bug based on the feature tests I'd written
Dependent destroy needed to be on the has-many side of the associations,
not on the belong_to side.

zporter commented on app/controllers/posts_controller.rb:
```diff
@@ -1,8 +1,10 @@
 class PostsController < ApplicationController
+  include HackernewsClone::VoteHelper
```
this appears to only be used in this controller. plans to expand in the future?

gingin77 replied:
yes - it'll be used by Comments

zporter commented on app/controllers/posts_controller.rb:
```diff
@@ -1,8 +1,10 @@
   before_action :authenticate_to_submit, only: :new
 
-  helper_method :post, :posts, :comment
+  helper_method :post, :posts, :comment, :new_vote
```
Could move this to the module like so:

```ruby
module VoteHelper
  def self.included(base)
    base.helper_method :new_vote
  end
end
```

zporter commented on app/controllers/votes_controller.rb:
```diff
@@ -1,2 +1,32 @@
+
+  def create
+    vote = current_user.votes.build(vote_params)
+    vote.save
```
I would skip creating the local var here.

gingin77 replied:
Thanks. I had an earlier version where I wanted some attribute from `vote` to inform the redirect path, but its not needed now.

zporter commented on app/controllers/votes_controller.rb:
```diff
@@ -1,2 +1,32 @@
+  end
+
+  def update
+    vote.value == 1 ? vote.value = -1 : vote.value = 1
```
I'm curious why the toggle approach here over passing a param to set the `value`?

gingin77 replied:
I guess its one less thing to pass in locals to the `shared/vote/edit` partial, but then I have a conditional here, so hmm... 🤔

zporter commented on app/models/post.rb:
```diff
@@ -29,6 +30,10 @@ def text_post?
 
+  def users_vote(user)
+    self.votes.find_by(user_id: user.id)
+  end
```
could this be a `has_many` on `User`?

gingin77 replied:
I'm not sure if this is meant to be taken at face value or is a leading question. I have now set up a `has_many(votes)` association on User. Were you suggesting I could rewrite this method as a feature of the User class? I use this method to evaluate how the vote buttons should be displayed depending on whether the current user has cast a vote for the particular post that's being evaluated. Since I'm going to see how I can simplify the conditionals I have in the views, I may not need this method, but wanted to ask about it in case there might be a better way to get the info I needed.

zporter replied:
I agree, I think we can re-evaluate this comment when you've revisited the views. I have a hunch that this can actually live on the `Vote` model, since this is trying to look up a specific Vote, but won't know for sure until the dust settles on the views. Let me know if you want to pair on those at all.

zporter commented on app/views/posts/_post_presenter_index.html.erb:
```diff
@@ -0,0 +1,9 @@
+    <span><%= link_to(post.title, post.url) %></span>
+  <% else %>
+    <span><%= post.text.truncate(80, {separator: " "}) %></span>
+  <% end %>
```
what if instead of a conditional, `post` had a `type` method that returned either `url` or `text` and that type was used to render a specific partial?  something like:

```erb
# _post_presenter_index.html.erb
<%= render partial: "#{post.type}_post_content", locals: { post: post } %>

# _url_post_content.html.erb
<span><%= link_to(post.title, post.url) %></span>

# _text_post_content.html.erb
<span><%= post.text.truncate(80, {separator: " "}) %></span>
```

A bit more flexible if more types were added in the future.

gingin77 replied:
I like that idea

zporter commented on app/views/posts/_post_presenter_show.html.erb:
```diff
@@ -30,6 +30,17 @@
         <td class="col2"><%= time_ago_in_words(post.created_at) %> ago</td>
       </tr>
+      <tr>
+        <td scope="row">vote count:</td>
```
`vote count` is a little misleading here. the value being printed out is more of a `score`.

gingin77 replied:
true

zporter commented on app/views/shared/vote/_active_button_options.html.erb:
```diff
@@ -0,0 +1,9 @@
+  <%= render partial: "shared/vote/up_to_down", locals: { parent: parent } %>
+<% else %>
+  <%= render partial: "shared/vote/down_to_up", locals: { parent: parent } %>
+<% end %>
```
these nested partials and conditionals are a little hard to follow. could the icon color be driven by a `disabled` state?  could `link_to_if` be used based on 1) whether the user is authenticated and 2) what their current vote value is for a given thing?

zporter commented on app/views/shared/vote/_edit.html.erb:
```diff
@@ -0,0 +1,5 @@
+<%= form_for(vote, url: vote_path(vote), action: "update") do |f| %>
+  <%= f.button type: 'submit', class: "my-button clickable", id: vote_type do %>
+    <%= tag.i class: "fa fa-chevron-circle-#{vote_type}" %>
```
what is `tag` here? I've never seen this before.

gingin77 replied:
tag is a rails helper for HTML tags - http://api.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-tag

I needed something to support the `<i>` that font awesome uses

zporter commented on app/views/shared/vote/_new.html.erb:
```diff
@@ -0,0 +1,8 @@
+<%= form_for(new_vote, url: votes_path, action: "create") do |f| %>
```
do `url` and `action` need to be set here?

zporter replied:
replace `new_vote` with `parent.votes.build(value: one)` and you can avoid setting the `value`s in the hidden inputs below.

zporter commented on spec/features/vote_on_posts_spec.rb:
```diff
@@ -0,0 +1,69 @@
+  let(:voter) { create(:oliver) }
+  let!(:url_post) { create(:url_post) }
+
+  before :each do
```
`:each` doesn't need to be specified here

zporter commented on spec/features/vote_on_posts_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    expect(page).to have_css "i.clickable, i.fa-times-circle"
+    expect(page).to have_link("delete", href: "/votes/#{Vote.last.id}")
+  end
+end
```
worth testing an unauthenticated user?

gingin77 replied:
yeah.... see https://github.com/gingin77/hackernews_clone/issues/29 "Sort out 2 issues related to voting feature test for non-auth user"

zporter commented on spec/models/post_spec.rb:
```diff
@@ -1,7 +1,8 @@
 RSpec.describe Post, type: :model do
-  it { should belong_to(:submitter).dependent(:destroy) }
+  let(:post) { build(:text_post) }
+  it { expect(post).to have_many(:comments).dependent(:destroy) }
```
nitpick: space between these 2 lines please

zporter replied:
line 5 could be `it { is_expected.to have_many(:comments).dependent(:destroy) }`

zporter commented on spec/models/voter_spec.rb:
```diff
@@ -2,9 +2,11 @@
-  it { should belong_to(:voter).dependent(:destroy) }
-  it { should belong_to(:voteable).dependent(:destroy) }
+  it { expect(submitter).to have_many(:votes).dependent(:destroy) }
+  it { expect(votable).to have_many(:votes).dependent(:destroy) }
```
I would expect these expectations to be with their respective models.

zporter reviewed (commented):
left some comments. Slack me when you've responded to them.

gingin77 committed:
Refactor in a few places

gingin77 committed:
Fix create method on votes controller

gingin77 committed:
Update Votes#create method; remove unneeded value fields from vote/new form; start new spec for non-auth voters

gingin77 committed:
WIP - refactoring methods in votes controller and forms

gingin77 committed:
Post-model_spec-remove unneeded let statement; update is_expected syntax

Update is_expected.to syntax for comment and user model specs

gingin77 committed:
Remove vote-helper module & method.

The method is not needed now, given the way the new-vote form is now set
up.

gingin77 committed:
New `type` method for Post class to use for view partial simplification

First, write model spec for method, then run integration tests to make
sure the views render as they need to.

gingin77 committed:
Remove flawed test for non-auth user tring to vote. Create issue #29.

zporter commented on app/controllers/votes_controller.rb:
```diff
@@ -1,2 +1,31 @@
+
+  def update
+    vote.value = vote_params[:value]
+    vote.save
```
could combine these 2 lines into `vote.update(value: vote_params[:value])`

gingin77 committed:
WIP - In middle of refactoring the way vote buttons are rendered

Current state is incomplete and inefficient, page reload times after
editing notes were - 4605ms, 4692ms, and 4716ms

gingin77 committed:
WIP - Posts#index page load times are currently around 100ms

gingin77 committed:
WIP - Define up_vote? and down_vote? predidate methods to disable button

Posts#index page load times are still around 100ms

gingin77 committed:
WIP - Refactor vote button views, functioning well, styles need fix

Page reload time stays ~100ms

gingin77 committed:
Update: Styles for vote buttons; vote view template for posts#show

Index page reload is ~100ms

gingin77 committed:
Create another test to evaluate dependent destroy for posts and votes

All tests are passing, except for one written to assess relationship between votes and comments

gingin77 committed:
Rename vote partials by dropping _buttons from name

gingin77 committed:
Set up relationship between Comments and Votes; add vote buttons to comment partial

All existing tests are currently passing

gingin77 committed:
Update seed file to create votes; remove unnessary local vars; in styles - adjust width for comment vote button div

gingin77 committed:
WIP - create CommentHelper module; refactor to remove redundant methods

gingin77 committed:
Refactor vote presenter view

gingin77 committed:
Refactor naming & organization of comment view partials & helper method

gingin77 committed:
Remove unused new_vote method from VoteHelper module

gingin77 committed:
Use div.vote_button_wrapper in the vote/presenter template

Remove same div wrapper from the static icons partial

gingin77 committed:
WIP: In the middle of refactoring code for new comment creation

A helper method that is used by both post and comment controllers has
been introduced - the necessary hidden fields have been added so that
that same partial can serve both parent types

Next, figure out how to put entire form in the partial, instead of
splitting form wrapper from form content

All existing tests are currently passing

gingin77 committed:
Refactor comment view partial relationships

Entire new comment form exists in one partial now, instead of being
split between parent and partial - more info is passed through locals,
including hte url path.

Next, evaluate whether 2 comment#create paths are essential

gingin77 committed:
WIP: Refactor new comment helpers, create paths, and views

Determination of commentable's class (ie Post vs Comment) no longer relies on params, though the id for each still is needed from params.

The parent of a new comment is now based on what is passed in locals to the "views/shared/comment/_form_new" partial.

Routes file was simplified because all comments#create actions can go through

After talking to Annie about how it was hard to keep track of helper methods used in views, I did further refactoring.

WIP Further refactoring of new comment creation

Remove parent.comments.build from the form view for a new comment.

Instead, initialize a new comment within the Posts and Comments controllers and pass these to views using helper methods.

Adjust helper method names as needed.

zporter commented on app/controllers/comments_controller.rb:
```diff
@@ -1,25 +1,24 @@
-    if @comment.save
-      redirect_to parent
+    @new_comment = current_user.comments.build(comment_params)
+    commentable = @new_comment.commentable
```
I'd use the `new_comment` reader method here and below since the ivar was set above.

zporter commented on app/views/posts/show.html.erb:
```diff
@@ -1,5 +1,9 @@
 <%= render partial: "post_presenter_show" %>
-<%= render partial: "shared/comment/form_new", locals: { comments_parent: post, url: (post_comments_path(post)), submit_text: "Comment On Post" }  %>
+<%= render partial: "shared/comment/form_new",
```
can this template live under `app/views/comments`?

gingin77 replied:
It could. I like having it under shared because both the Posts and Comments controllers are using it, but if having a shared folder is seen as an excess, then I'd want to move both comment partials to `views/comments`, and then move everything from `shared/votes` to a `votes` dir. My thinking was that any partials that are being used by another controller would go in the `shared` dir.

zporter replied:
I can dig that line of thinking 👍. I just figured the form is related to comments, so it should live under the comments dir.

gingin77 committed:
Add CommentHelper module to def commentable for all controllers

In form for new comment:
- inline `commentable.comments.build`
- remove `url` and `method` options - they are not needed
- remove values on hidden fields, as they are also not needed due to
inlining `commentable.comments.build`

Update Comment#create method, include the commenthelper module in both
the Posts and Comments controllers

gingin77 committed:
Remove modules wrapped around variables for content files in seeds dir

gingin77 committed:
Add erd gem to generate a diagram of relationships between models

zporter commented on app/controllers/comments_controller.rb:
```diff
@@ -1,21 +1,25 @@
 class CommentsController < ApplicationController
-  before_action :authenticate_to_submit, only: :new
+  include HackernewsClone::VoteHelper
+  include HackernewsClone::CommentHelper
```
I think this is where `app/controllers/concerns` can step in to help out. something like `app/controllers/concerns/voteable_controller.rb` and `app/controllers/concerns/commentable_controller.rb` defining `VoteableController` and `CommentableController` respectively would be more inline with the Rails way™ and move away from the `Helper` suffix which might be confused with files in the `app/helpers` dir.

gingin77 replied:
I made this change, then I thought about things more and I don't think the VoteHelper method belongs in a concerns dir under controllers. The method is really a query method that looks up whether a vote exists which belongs to the current user and the parent instance that a given vote button is attached to. I'm working on how to write the code so that the parent and current user get passed from the view to the vote_presenter.

gingin77 committed:
Mv commentable and voteable helper mods from lib to controllers/concerns

Modify the Posts and Comments controllers as necessary to reflect the new file paths and names

gingin77 committed:
WIP Move module with parent_vote(parent) back to lib.

I think I want this to be a view helper, rather than something in the
controller concerns - or it should be part of the model or in a querry
object.

I made some changes to the application helper in an effort to implement
the approach that Ryan Bates took in his postcast on presenters.

zporter commented on app/controllers/concerns/voteable_controller.rb:
```diff
@@ -0,0 +1,11 @@
+  end
+
+  def parents_vote(parent)
+    @vote = Vote.find_by(user_id: current_user, voteable_id: parent)
```
unnecessarily setting an ivar here. I'd also probably find some way to put a method on `Vote` to having something like this in your views `current_user.votes.find_by(voteable: parent)`. there might be more performant ways to go about this, but this should get the job done without requiring a method on the controllers.

gingin77 replied:
cool - thanks - closing this PR - have copied this comment to Issue https://github.com/gingin77/hackernews_clone/issues/30
